test/extended/baremetal/high_availability.go

```diff
@@ -0,0 +1,160 @@
+				if errors.IsNotFound(err) {
+					return false, nil
+				}
+
+				e2e.Logf("Error getting cluster baremetal operator deployment: %v", err)
+				return false, nil
+			}
+
+			if dc.Status.AvailableReplicas != 1 {
+				return false, nil
+			}
+
+			baremetalCO, err := oc.AdminConfigClient().ConfigV1().ClusterOperators().Get(context.Background(), "baremetal", v1.GetOptions{})
+			if err != nil {
+				e2e.Logf("Error getting authentication operator: %v", err)
+				return false, nil
+			}
+
+			if clusteroperatorhelpers.IsStatusConditionFalse(baremetalCO.Status.Conditions, configv1.OperatorAvailable) {
+				return false, nil
+			}
+
+			return true, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("delete metal3 deployment")
+		err = c.AppsV1().Deployments(baremetalNamespace).Delete(context.Background(), metal3Deployment, v1.DeleteOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("wait until metal3 deployment is deleted")
+		err = wait.PollImmediate(deleteInterval, deleteWaitTimeout, func() (bool, error) {
+			_, err := c.AppsV1().Deployments(baremetalNamespace).Get(context.Background(), metal3Deployment, v1.GetOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					return true, nil
+				}
+
+				e2e.Logf("Error getting metal3 deployment: %v", err)
+				return false, nil
+			}
+
+			return false, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("wait until metal3 deployment returns back healthy")
+		err = wait.Poll(restartInterval, restartWaitTimeout, func() (bool, error) {
+			dc, err := c.AppsV1().Deployments(baremetalNamespace).Get(context.Background(), metal3Deployment, v1.GetOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					return false, nil
+				}
+
+				e2e.Logf("Error getting metal3 deployment: %v", err)
+				return false, nil
+			}
+
+			if dc.Status.AvailableReplicas != 1 {
+				return false, nil
+			}
+
+			return true, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("verify that baremetal hosts are healthy and correct state")
+		checkMetal3DeploymentHealthy(oc)
+	})
+})
+
+func checkMetal3DeploymentHealthy(oc *exutil.CLI) {
+	dc := oc.AdminDynamicClient()
+	bmc := baremetalClient(dc)
+
+	hosts, err := bmc.List(context.Background(), v1.ListOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(hosts.Items).ToNot(o.BeEmpty())
+
+	for _, h := range hosts.Items {
+		expectStringField(h, "baremetalhost", "status.operationalStatus").To(o.BeEquivalentTo("OK"))
+		if strings.Contains(h.GetName(), "master") {
+			expectStringField(h, "baremetalhost", "status.provisioning.state").To(o.BeEquivalentTo("externally provisioned"))
+			expectBoolField(h, "baremetalhost", "spec.online").To(o.BeTrue())
+		} else {
+			// extra workers are in available status and online field is set to false
+			expectStringField(h, "baremetalhost", "status.provisioning.state").To(
+				o.Or(
+					o.BeEquivalentTo("provisioned"),
+					o.BeEquivalentTo("available")))
```

Comment:
What's preventing using the same code block used in this [test](https://github.com/openshift/origin/blob/f1cfa98d3dd57179844c0acad5f9fb673c965673/test/extended/baremetal/hosts.go#L136-L146) to verify that the metal3 resources are in a good shape?

Reply:
The test you referenced is parallel suite test and there is no extra worker. We can safely ensure that every host is provisioned and online. However, this test is serial, there is an extra worker(status: available, online:false).

Reply:
But after our slack discussions, I started thinking that maybe we may not check worker's statuses and even we can remove all status check. What do you think?

Reply:
Here the provisioning state check is fundamental, since it ensures that the restart of the metal3 deployment didn't affect the (already existing) baremetal resources. The main problem is related due the extraworker management, which anyhow we're going to change (thus the logic will become the same of the other tests, ie checking that everything is provisioned or externally provisioned))

Reply:
Yes, agree. This PR can wait until extra worker management is finalized. After that, this check will be unified with the parallel test check.